euicc/hexutil.c

```diff
@@ -1,142 +1,110 @@
-        } else {
-            return -1;
-        }
-
-        output[i] = (high << 4) + low;
+    memcpy(bin, input, n);
+    memset(bin + n, 'f', n % 2); // pad with 'f' if odd
+    const int bytes = euicc_hexutil_hex2bin_r(output, output_len, bin, n + (n % 2));
+    free(bin);
+    if (bytes == -1) {
+        return -1;
     }
-
-    return length;
+    for (int i = 0; i < bytes; i++) {
+        output[i] = output[i] >> 4 | (output[i] & 0xf) << 4; // swap
+    }
+    memset(output + bytes, 0xff, padding_to - bytes);
+    return bytes;
 }
 
-int euicc_hexutil_gsmbcd2bin(uint8_t *output, uint32_t output_len, const char *str, uint32_t padding_to) {
-    uint32_t str_length;
-    uint32_t idx = 0;
-
-    str_length = strlen(str);
-
-    if (output_len < (str_length + 1) / 2) {
+int euicc_hexutil_bin2gsmbcd(char *output, const uint32_t output_len, const uint8_t *bin, const uint32_t bin_len) {
+    if (output == NULL || bin == NULL || output_len < ((2 * bin_len) + 1)) {
         return -1;
     }
-
-    if (output_len < padding_to) {
+    uint32_t i;
+    uint8_t *tmp = calloc(bin_len, sizeof(uint8_t));
+    if (tmp == NULL) {
         return -1;
     }
-
-    for (uint32_t i = 0; i < str_length; i += 2) {
-        char high_nibble = (i + 1 < str_length) ? str[i + 1] : 'F';
-        char low_nibble = str[i];
-
-        uint8_t high_nibble_val = 0x0;
-        uint8_t low_nibble_val = 0x0;
-
-        if (low_nibble >= '0' && low_nibble <= '9') {
-            low_nibble_val = low_nibble - '0';
-        } else if (low_nibble == 'F' || low_nibble == 'f') {
-            low_nibble_val = 0x0F;
-        } else {
-            return -1;
-        }
-
-        if (high_nibble >= '0' && high_nibble <= '9') {
-            high_nibble_val = high_nibble - '0';
-        } else if (high_nibble == 'F' || high_nibble == 'f') {
-            high_nibble_val = 0xF;
-        } else {
-            return -1;
-        }
-
-        output[idx] = (high_nibble_val << 4) | low_nibble_val;
-        idx++;
-    }
-
-    for (; idx < padding_to; idx++) {
-        output[idx] = 0xFF;
+    for (i = 0; i < bin_len; i++) {
+        tmp[i] = bin[i] >> 4 | (bin[i] & 0xf) << 4; // swap
     }
-
-    return idx;
-}
-
-int euicc_hexutil_bin2gsmbcd(char *output, uint32_t output_len, const uint8_t *binData, uint32_t length) {
-    if (euicc_hexutil_bin2hex(output, output_len, binData, length)) {
+    const int ret = euicc_hexutil_bin2hex(output, output_len, tmp, bin_len);
+    free(tmp);
+    if (ret != 0) {
         return -1;
     }
-
-    length = strlen(output);
-    for (size_t i = 0; i < length - 1; i += 2) {
-        char temp = output[i];
-        output[i] = output[i + 1];
-        output[i + 1] = temp;
-    }
-
-    for (uint32_t i = length - 1; i > 0; i--) {
-        if (output[i] != 'f') {
-            break;
-        }
-        output[i] = '\0';
+    i = (2 * bin_len) - 1;
+    while (i > 0 && output[i] != 'f') {
+        i--;
     }
-
+    output[i] = 0;
```

Comment:
```suggestion
    output[i] = '\0';
```
Use '\0' for readability.